### PR TITLE
copy user feature for older liberty versions

### PR DESCRIPTION
**Pull request by jjiwooLim**
https://github.com/OpenLiberty/ci.maven/issues/1532
Copy user feature jar and manifest file to Liberty image when installFeatureUtility/featureUtility is not available (for Liberty 21.0.0.11 or below).

### CODE REVIEW

File: src/main/java/io/openliberty/tools/common/plugins/util/InstallFeatureUtil.java [hunk 1]
```diff
@@ -29,13 +29,17 @@
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.URLClassLoader;
+import java.nio.file.FileSystems;
```
Comment (cherylking) [inline src/main/java/io/openliberty/tools/common/plugins/util/InstallFeatureUtil.java:32]: Can you update the copyright year from 2021 to 2023?

File: src/main/java/io/openliberty/tools/common/plugins/util/InstallFeatureUtil.java [hunk 2]
```diff
@@ -527,12 +555,65 @@ private static Set<String> toLowerCase(Collection<String> strings) {
...
+		            		shortNameMap.put(symbolicName.toLowerCase(), "");
+		            	}
+		            	File targetFile = new File(featuresDirectory, symbolicName + ".mf");
+		            	if(targetFile.exists()) {
+		            		info("The feature " + esa + " is already installed.");
+		            		break;
+		            	}
+		            	copyZipEntry(zip, entry, targetFile);
+		            	
+		            }else if(fileName.toLowerCase().endsWith("jar")){
```
Comment (cherylking) [inline src/main/java/io/openliberty/tools/common/plugins/util/InstallFeatureUtil.java:596]: Should compare to `.jar` not `jar` to be safe.

File: src/main/java/io/openliberty/tools/common/plugins/util/InstallFeatureUtil.java [hunk 3]
```diff
@@ -545,17 +626,41 @@ public void installFeatures(boolean isAcceptLicense, List<String> featuresList)
...
+    	DefaultArtifactVersion version;
+    	if(openLibertyVersion != null) {
+    		version = new DefaultArtifactVersion(openLibertyVersion);	
+    		info("plugin listed esa: " + pluginListedEsas.toString());
+    		if (version.compareTo(MIN_VERSION) < 0 && !pluginListedEsas.isEmpty()) {
+    			//manually install user feature esas
+    			info("Neither InstallUtility nor FeatureUtility is available to install user feature esa.");
+    			info("Attempting to manually install the user feature esa without resolving its dependencies.");
+    			info("Recommended user action: upgrade to OpenLiberty version " + MIN_USER_FEATURE_VERSION + " or higher and provide features-bom file for the user feature esa.");
+    			info(" To directly install the esa file without providing features-bom file, upgrade to OpenLiberty version 23.0.0.2.");
```
Comment (cherylking) [inline src/main/java/io/openliberty/tools/common/plugins/util/InstallFeatureUtil.java:639]: Include `23.0.0.2 or later.` in this log message.

File: src/main/java/io/openliberty/tools/common/plugins/util/InstallFeatureUtil.java [hunk 4]
```diff
@@ -545,17 +626,41 @@ public void installFeatures(boolean isAcceptLicense, List<String> featuresList)
...
+    		   	
     	for (String feature: featuresList) {
     		if (feature.contains(":")) {
     			String[] userFeatureSplit = feature.split(":");
-    			featureToExtMap.put(userFeatureSplit[1], userFeatureSplit[0]);
-    			featuresToInstall.add(userFeatureSplit[1]);
+    			String userFeatureName = userFeatureSplit[1].toLowerCase();
+    			featureToExtMap.put(userFeatureName, userFeatureSplit[0]);
+    			//check if user feature esa was installed manually
+    			if(!shortNameMap.containsValue(userFeatureName) && !shortNameMap.containsKey(userFeatureName)) {
```
Comment (cherylking) [inline src/main/java/io/openliberty/tools/common/plugins/util/InstallFeatureUtil.java:651]: I don't understand why it is checking for the feature name in both the key and value of `shortNameMap`. That map uses the symbolic name for the key. How does this `userFeatureName` map to the `Subsystem-SymbolicName` and the `IBM-ShortName` referenced when populating the `shortNameMap`?

**Review by cherylking — CHANGES_REQUESTED** [review]

File: src/main/java/io/openliberty/tools/common/plugins/util/InstallFeatureUtil.java [hunk 5]
```diff
@@ -159,13 +180,29 @@ public InstallFeatureUtil(File installDirectory, File buildDirectory, String fro
...
+	private void copyZipEntry(ZipFile zip, ZipEntry entry, File targetFile) throws IOException, FileNotFoundException {
+		try(InputStream is = zip.getInputStream(entry)){	
+			// write file content
+		    try(FileOutputStream fos = new FileOutputStream(targetFile)){
+		    	int len;
+		    	byte[] buffer = new byte[1024];
+		        while ((len = is.read(buffer)) > 0) {
+		            fos.write(buffer, 0, len);
+		        }
+		    }
```
[the remark this answers is not in the document]
Comment (jjiwooLim) [reply src/main/java/io/openliberty/tools/common/plugins/util/InstallFeatureUtil.java:204]: I am using try-with-resources statement here. This statement ensures that each resource is closed at the end of the statement so `finally` is not needed.

**Review by cherylking — APPROVED** [review]

**Pull request merged**